### PR TITLE
Add preview and replace button to downloads edit

### LINKED ISSUE #35143: [Downloads] Edit file modal
**Job story**

- When I put my digital product or service on sale on my website,
- I want to manage basic file settings
- So that the customer has a better post-purchase experience

**Design**

AkVNGImLgSqCObTQ3idVn7-fi-7001_309093

**Acceptance criteria**

- [ ] The user can click the Edit button to open a model with file details from each file row.
- [ ] The modal has the file’s name (custom or original) in the title.
- [ ] We display a file thumbnail next to its name. If we can’t obtain it, we show a grey rectangle.
- [ ] The user can swap the file for another one using the Replace button. The original file is removed from the list (but not from the library), and the new one is added in its place. The new file maintains the old one’s name.
- [ ] Clicking the Replace button opens the WordPress upload modal.
- [ ] The user can change the file name. It will be shown to customers on the order confirmation page and email. The name is only a label and doesn’t replace the file name.
- [ ] The file URL is displayed in a disabled, non-interactive text field. The user can copy the URL either by highlighting the field’s contents or clicking the blue icon on the right side.
- [ ] When the user clicks Save, we close the modal and immediately update the list. If needed, we show the loading spinner to the button, then close the modal.
- [ ] When the user clicks "Delete file", we remove the file without any confirmation. We're not deleting it from the library, just from the product.

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR is a follow-up to [PR 40599](https://github.com/woocommerce/woocommerce/pull/40599). This PR adds a preview of the uploaded file and a replace button [that opens the system file picker](https://github.com/woocommerce/woocommerce/issues/35143#issuecomment-1770212954).

Screenshot 2023-10-20 at 15 33 46

Closes #35143.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a WooCommerce store with this branch and enable the product editor and the virtual-downloadable products (`product-virtual-downloadable`) using **WooCommerce > Settings > Advanced > Features**
2. Go to **Products > Add New**
3. In the `General` tab, scroll to the bottom of the page where a new `Downloads` section should appear.
4. By default, the list displays files with their actual names, including the extension/format, after uploading one or more files.
Each file is listed as a separate item with options to remove (using `X`) and `Edit` them.
6. Click the `Edit` button. A modal will be visible.
7. If the uploaded file was an image, confirm that the modal displays it.
8. Verify the `Replace` button works as expected.
9. After replacing a file, it should be loaded in the modal and updated in the list after closing the modal.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss [hunk 1]
```diff
@@ -15,7 +15,26 @@
...
 	}
+
+	.woocommerce-image-gallery, .components-form-file-upload {
+		display: inline-block;
+		vertical-align: middle;
+		p {
+			margin-bottom: $gap-smallest;
+		}
+		button {
+			color: $gutenberg-blue;
```
Comment (louwie17) [inline packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss:37]: This doesn't exist, I noticed you created this before in another css file, so maybe do something similar?
`$gutenberg-blue: var(--wp-admin-theme-color);`?

**Review by louwie17 — CHANGES_REQUESTED** [review]
Overall this is testing well and code looks good. I did leave an inline comment about a build error.
Also the preview seems a bit on the big side compared to the wireframes. The one now has a max height of 146px but the wireframe only has 73/74px for the image.
Screenshot 2023-10-23 at 10 37 25 AM
versus:
Screenshot 2023-10-23 at 10 41 18 AM

**Comment by octaedro** [comment]
Thank you @louwie17 for your review.
I've just fixed the error you mentioned and resized the image. I stuck to the original design and set it to 72px.
Could you take another look at this PR?

File: packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss (hunk 1 above)
Comment (octaedro) [reply packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss:37]: Good call @louwie17! I fixed it in the commit [475cc59](https://github.com/woocommerce/woocommerce/pull/40835/commits/475cc59847837dd63a8cc963c26811c513962b0e)

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for adjusting the preview size, the smaller size looks much better.
There is just one other minor styling issue, it seems like the spacing between the preview and text is a lot bigger then it should be ( 48px vs 16px ).
Seems like the gap does something weird for the image wrapper:
Screenshot 2023-11-02 at 11 44 11 AM

**Comment by octaedro** [comment]
Good eyes @louwie17! The gap was because by default the image gallery shows 4 columns, I just fixed it in the commit [6718304](https://github.com/woocommerce/woocommerce/pull/40835/commits/671830433f5097d37153bb52d20fbab7d4cca2a4).
Could you give this PR another look?

File: packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss [hunk 2]
```diff
@@ -25,9 +25,11 @@ $gutenberg-blue: var(--wp-admin-theme-color);
 		cursor: pointer;
 	}
 
-	.woocommerce-image-gallery, .components-form-file-upload {
+	.woocommerce-image-gallery,
+	.components-form-file-upload {
 		display: inline-block;
 		vertical-align: middle;
+		padding-left: $gap;
```
Comment (louwie17) [inline packages/js/product-editor/src/blocks/product-fields/downloads/edit-downloads-modal/style.scss]: Adding the padding here means its also being applied to the `woocommerce-image-gallery` and thus causing the image to also move to the right, could we only add the padding to the `components-form-file-upload`?
Screenshot 2023-11-03 at 1 17 57 PM

**Review by louwie17 — CHANGES_REQUESTED** [review]
@octaedro I hate to be the bearer of bad news haha I just found another style issue 🙈 
Hopefully this will be the last one.

**Review by louwie17 — APPROVED** [review]
Thanks for the several updates @octaedro, things are looking great now and testing well, nice work!!

**Pull request merged**